### DISCUSSION TITLE
VFO ideas

### DISCUSSION THREAD

**Discussion by awandahl** [mbridak/not1mm#207] [Ideas]
I have played around with the vfo. I let a friend print the box and he added the front plate too. Code will be supplied for that later. When I made holes for the encoder I also drilled holes for a switch and a two color led. I wasn't sure what I wanted to make with the switch/leds at that time but I thought it was good to be prepared for something. I have also tried to make use of the switch by changing the speed of dialing, introducing a fine (original) and a more coarse mode. It kind of works but the latency is annoying (of course). So what I want to discuss is how latency possibly can be diminished. Higher speed in the communication between vfo and not1mm? What implications can that have? And if I have a switch, what else could be controlled, more useful than controlling fine/coarse dialling? The code which I have got from some iterations at perplexity.ai  How can the code be improved?
```
import rotaryio
import time
from board import *
import digitalio
import usb_cdc

led = digitalio.DigitalInOut(LED)
led.direction = digitalio.Direction.OUTPUT

green_led = digitalio.DigitalInOut(GP19)
green_led.direction = digitalio.Direction.OUTPUT

red_led = digitalio.DigitalInOut(GP18)
red_led.direction = digitalio.Direction.OUTPUT

switch = digitalio.DigitalInOut(GP20)
switch.direction = digitalio.Direction.INPUT
switch.pull = digitalio.Pull.UP

serial = usb_cdc.console

enc = rotaryio.IncrementalEncoder(GP22, GP21)
enc.position = 0

last_position = None
inputstring = ""
fine_mode = True  # Start in fine mode

def out(data):
    usb_cdc.console.write(data.encode())

def update_leds():
    green_led.value = fine_mode
    red_led.value = not fine_mode

update_leds()  # Initialize LEDs

last_switch_state = switch.value
last_toggle_time = time.monotonic()

while True:
    current_time = time.monotonic()
    current_switch_state = switch.value

    if current_switch_state != last_switch_state and (current_time - last_toggle_time) > 0.2:
        if current_switch_state == False:  # Switch is pressed (active low)
            fine_mode = not fine_mode
            update_leds()
            last_toggle_time = current_time
        last_switch_state = current_switch_state

    current_position = enc.position
    if last_position is None:
        last_position = current_position
    
    if current_position != last_position:
        change = current_position - last_position
        if fine_mode:
            change = change // 20  # Finer steps
        else:
            change = change // 1  # Coarser steps
        
        if change != 0:
            enc.position = last_position + (change * (1 if fine_mode else 20))
            out(f"{enc.position}\r\n")
        last_position = enc.position

    if usb_cdc.console.in_waiting:
        led.value = True
        incoming = usb_cdc.console.read(usb_cdc.console.in_waiting)
        inputstring = inputstring + incoming.decode()

        if inputstring[-1] == "\r":
            if inputstring.strip() == 'f':
                position = enc.position
                out(f"{position}\r\n")
            if inputstring[0] == "F" and len(inputstring.strip().split()) == 2:
                freq = inputstring.strip().split()[1]
                if freq.isdigit():
                    enc.position = int(freq)
                    last_position = enc.position
            if inputstring.strip() == 'whatareyou':
                out("vfoknob\r\n")
            inputstring = ""
        led.value = False

    time.sleep(0.01)  # Small delay to prevent busy-waiting
```

**Comment by awandahl** [comment]
![IMG_6350](https://github.com/user-attachments/assets/c19d3588-2d3f-456a-93f0-a2a5d7eee999)
![IMG_6349](https://github.com/user-attachments/assets/7ad997ad-13ac-4d28-b4f6-d431683aba75)

**Comment by mbridak** [comment]
Hey, Nice looking build.

It's been awhile since I had messed with this. If I remember correctly one thing I was fighting that I wanted to fix was that as you're tuning the knob you're advancing the internal state and about once a second the poll of the physical radio freq is sent to the knob. This causes kind of a jump back in the knob freq. Kind of a two steps forward, one step back.

I believe what I was thinking about doing was having somesort of a delta timeout before the know would accept updated from the radio. This was the internal frequency state of the knob would continue advancing forward. and say like after a second of the knob not rotating it would start accepting changes from the radio again.

Button usage ideas...
- Tune step multiplier.
- Toggle the knob to control the bandwidth.
- Same for RIT

**Reply by awandahl** [reply]
RIT on/off would be more useful than step multiplier actually. Thanks!
